### PR TITLE
Update migration guide /8

### LINKED ISSUE #21570: 6.0 Migrate from ASP.NET to ASP.NET Core
Update to ~3.1~ ***UPDATED***  6.0 when 6.0 releases and 2.1 is dropped from LTS

The entire series needs to be review by the PU
Review  #16437 and #14386

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 1133d6d1-7801-654c-2b7c-dad2bb342501
* Version Independent ID: a7ccce22-f905-06f4-8847-b9eeaf534e1f
* Content: [Migrate from ASP.NET to ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/migration/proper-to-2x/?view=aspnetcore-5.0)
* Content Source: [aspnetcore/migration/proper-to-2x/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/migration/proper-to-2x/index.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-migration**
* GitHub Login: @isaac2004
* Microsoft Alias: **scaddie**

---
[Associated WorkItem - 60622](https://dev.azure.com/msft-skilling/Content/_workitems/edit/60622)

### LINKED ISSUE #21966: Consolidate ASP.NET to ASP.NET Core migration guidance
In addition to this section on migrating from ASP.NET to ASP.NET Core, we also have an eBook on the topic at https://docs.microsoft.com/dotnet/architecture/porting-existing-aspnet-apps/. We should consolidate all of this migration content into the eBook and then make the "Migration > ASP.NET to ASP.NET Core" node in the ASP.NET Core docs point to the eBook.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 1133d6d1-7801-654c-2b7c-dad2bb342501
* Version Independent ID: a7ccce22-f905-06f4-8847-b9eeaf534e1f
* Content: [Migrate from ASP.NET to ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/migration/proper-to-2x/?view=aspnetcore-5.0)
* Content Source: [aspnetcore/migration/proper-to-2x/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/migration/proper-to-2x/index.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-migration**
* GitHub Login: @isaac2004
* Microsoft Alias: **scaddie**

---
[Associated WorkItem - 60643](https://dev.azure.com/msft-skilling/Content/_workitems/edit/60643)

### LINKED ISSUE #22352: Doc URI differences between ASP.NET and Core
ASP.NET Core has some URI decoding differences compared to ASP.NET 4.x. Here's a recent Q&A:

> The only differences I’m seeing are 1) %5C (i.e., \ ) gets converted to a / in ASP.NET but is respected as a \ in ASP.NET Core and 
 
The %5C conversion to / was a deviation from the spec in IIS that we’ve avoided in Core. It was derived from the outdated idea that all of your paths were based on file system paths and it tried to mitigate some spoofing situations.

> 2) %2F (i.e., / ) gets properly converted to a / in ASP.NET but gets left as %2F in ASP.NET Core.

This one we don’t handle as gracefully. The whole path gets un-escaped except %2F because converting it to / would change the path structure. It can’t be decoded until the path is split into segments.

> For the HttpRequest.Url, I’m returning “new Uri(this.AspNetCoreHttpRequest.GetDisplayUrl());”. GetDisplayUrl returns a url which has the same quirks as above but otherwise is correct. The problem is when I wrap it in a Uri object, the Uri class does some decoding (e.g. “/’%20’” turns into “/’ ‘”) which doesn’t match the behavior in ASP.NET. 

That should be “new Uri(this.AspNetCoreHttpRequest.GetEncodedUrl());” to avoid Uri misinterpreting the values.

@davidfowl 
---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 1133d6d1-7801-654c-2b7c-dad2bb342501
* Version Independent ID: a7ccce22-f905-06f4-8847-b9eeaf534e1f
* Content: [Migrate from ASP.NET to ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/migration/proper-to-2x/?view=aspnetcore-5.0)
* Content Source: [aspnetcore/migration/proper-to-2x/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/migration/proper-to-2x/index.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-migration**
* GitHub Login: @isaac2004
* Microsoft Alias: **scaddie**

---
[Associated WorkItem - 60645](https://dev.azure.com/msft-skilling/Content/_workitems/edit/60645)

### LINKED ISSUE #27611: User Secrets documentation doesn't mention incompatibility with AssemblyInfo.cs
`AddUserSecrets<T>()` depends on the presence of the `UserSecretsIdAttribute` decorating the calling assembly. This documentation only mentions the **project setting** and doesn't clarify that what this does is generate an assembly attribute, and in particular, this doesn't happen if the project was migrated from a previous version of .NET using the Migration Assistant, where .NET Framework project conventions were typically to use an `AssemblyInfo.cs` and therefore migration adds a project setting `<GenerateAssemblyInfo>false</GenerateAssemblyInfo>` since generated attributes and boilerplate coded attributes conflict. This setting indirectly and silently prevents user secrets from working. It's particularly obtuse in .NET 6 and onwards because the AddUserSecrets call is implicit in the default builder and you wouldn't think look at the method signature and deduce what it's probably doing. Passing `optional: false` exposes the error more clearly, but again if this is implicit in the default builder, one wouldn't think to set it.

I don't think it's practical to change the API behaviour since user secrets should be an opt-in, but when one does opt-in on a migrated project, it simply doesn't work unless you know to also change something that would seem unrelated. Therefore I think this needs to be documented more clearly.
 
---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 9c99993c-c983-6fda-6c3a-ce4ec8c1d5db
* Version Independent ID: ba688ac2-e90e-fce5-f60d-40c2553c3efd
* Content: [Safe storage of app secrets in development in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-5.0&tabs=windows#register-the-user-secrets-configuration-source)
* Content Source: [aspnetcore/security/app-secrets.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/security/app-secrets.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-security**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

---
[Associated WorkItem - 60873](https://dev.azure.com/msft-skilling/Content/_workitems/edit/60873)

### LINKED ISSUE #27786: .NET 7: Update "Migrate from ASP.NET to ASP.NET Core" and reconcile with incremental
Per https://github.com/dotnet/systemweb-adapters/issues/98

Reconcile this doc with the [Incremental ASP.NET to ASP.NET Core migration](https://learn.microsoft.com/en-us/aspnet/core/migration/inc/overview) series.

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 1133d6d1-7801-654c-2b7c-dad2bb342501
* Version Independent ID: a7ccce22-f905-06f4-8847-b9eeaf534e1f
* Content: [Migrate from ASP.NET to ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/migration/proper-to-2x/?view=aspnetcore-7.0)
* Content Source: [aspnetcore/migration/proper-to-2x/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/migration/proper-to-2x/index.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-migration**
* GitHub Login: @isaacrlevin
* Microsoft Alias: **riande**

---
[Associated WorkItem - 60644](https://dev.azure.com/msft-skilling/Content/_workitems/edit/60644)

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #21966 Fixes #27786 Fixes #22352 Fixes #21570 Fixes #27611

[Internal review ULR](https://review.learn.microsoft.com/en-us/aspnet/core/migration/proper-to-2x/?view=aspnetcore-8.0&branch=pr-en-us-28184)

### CODE REVIEW

File: aspnetcore/migration/proper-to-2x/index.md [hunk 1]
```diff
@@ -1,207 +1,39 @@
...
-    ├── ...
-    └── web.config
-```
-
-## [BIND] and Input Formatters
-
-[Previous versions of ASP.NET](/aspnet/mvc/overview/getting-started/introduction/examining-the-edit-methods-and-edit-view) used the `[Bind]` attribute to protect against overposting attacks. [Input formatters](xref:mvc/models/model-binding#input-formatters) work differently in ASP.NET Core. The `[Bind]` attribute is no longer designed to prevent overposting when used with input formatters to parse JSON or XML. These attributes affect model binding when the source of data is form data posted with the `x-www-form-urlencoded` content type.
-
-For apps that post JSON information to controllers and use JSON Input Formatters to parse the data, we recommend replacing the `[Bind]` attribute with a view model that matches the properties defined by the `[Bind]` attribute.
+To generate the value for `HttpRequest.Url`, use `new Uri(this.AspNetCoreHttpRequest.GetEncodedUrl());` to avoid `Uri` misinterpreting the values.
```
Comment (Rick-Anderson) [inline aspnetcore/migration/proper-to-2x/index.md:36]: @GREsau @RealSwimburger %loop-evgeny please review this section.
See #22352

File: aspnetcore/migration/proper-to-2x/index.md [hunk 2]
```diff
@@ -1,211 +1,47 @@
...
 
-## Target frameworks
+When decoding `%2F` on ASP.NET Core:
 
-ASP.NET Core projects offer developers the flexibility of targeting .NET Core, .NET Framework, or both. See [Choosing between .NET Core and .NET Framework for server apps](/dotnet/standard/choosing-core-framework-server) to determine which target framework is most appropriate.
+* The entire path gets unescaped except `%2F` because converting it to `/` would change the path structure. It can’t be decoded until the path is split into segments.
+* See the [PathBase source](https://source.dot.net/#Microsoft.AspNetCore.Http.Abstractions/HttpRequest.cs,8d85f458c32cb4a5) for more information.
 
-When targeting .NET Framework, projects need to reference individual NuGet packages.
+To generate the value for `HttpRequest.Url`, use `new Uri(this.AspNetCoreHttpRequest.GetEncodedUrl());` to avoid `Uri` misinterpreting the values.
```
Comment (Rick-Anderson) [inline aspnetcore/migration/proper-to-2x/index.md:36]: @Tratcher please review. This section fixes #22352 and doesn't seem to fit in the incremental docs.

File: aspnetcore/migration/proper-to-2x/index.md [hunk 3]
```diff
@@ -1,211 +1,47 @@
...
-See the ebook [Porting existing ASP.NET apps to .NET Core](https://aka.ms/aspnet-porting-ebook) for a comprehensive porting guide.
+## URI decoding differences between ASP.NET to ASP.NET Core
 
-## Prerequisites
+ASP.NET Core has the following URI decoding differences with ASP.NET Framework:
 
-[.NET Core SDK 2.2 or later](https://dotnet.microsoft.com/download)
+| ASCII   | Encoded | ASP.NET Core | ASP.NET Framework |
+| ------------- | ------------- | ------------- | ------------- |
+| `\` | `%5C`  |  `/` |  `\` |
```
Comment (Tratcher) [inline aspnetcore/migration/proper-to-2x/index.md:27]:
```suggestion
| `\` | `%5C`  |  `\` |  `/` |
```

File: aspnetcore/migration/proper-to-2x/index.md [hunk 4]
```diff
@@ -1,211 +1,47 @@
...
+| `/` | `%2F`  |  `%2F` |  `/` |
 
-## Target frameworks
+When decoding `%2F` on ASP.NET Core:
 
-ASP.NET Core projects offer developers the flexibility of targeting .NET Core, .NET Framework, or both. See [Choosing between .NET Core and .NET Framework for server apps](/dotnet/standard/choosing-core-framework-server) to determine which target framework is most appropriate.
+* The entire path gets unescaped except `%2F` because converting it to `/` would change the path structure. It can’t be decoded until the path is split into segments.
+* See the [PathBase source](https://source.dot.net/#Microsoft.AspNetCore.Http.Abstractions/HttpRequest.cs,8d85f458c32cb4a5) for more information.
```
Comment (Tratcher) [inline aspnetcore/migration/proper-to-2x/index.md:33]:
```suggestion
```

File: aspnetcore/migration/proper-to-2x/index.md (hunk 2 above)
Comment (Tratcher) [reply aspnetcore/migration/proper-to-2x/index.md:35]: :shipit:

**Pull request merged**